File: apps/Applications/ios/Local For Vocal/Local For Vocal/Sources/Managers/ImageDiskCache.swift
import CryptoKit
import Foundation
import UIKit

/// Persistent image disk cache for offline support
/// Stores downloaded images to Library/Caches/Images/ with SHA256-hashed filenames
/// Thread-safe via actor isolation
actor ImageDiskCache {
    static let shared = ImageDiskCache()

    private let fileManager = FileManager.default
    private let cacheDirectory: URL
    private let maxCacheSize: Int64 = 500 * 1024 * 1024  // 500 MB

    // In-memory index for fast lookups: URL hash → file exists
    private var fileIndex: Set<String> = []

    // Fix #13: Track total cache size incrementally
    private var currentCacheSize: Int64 = 0

    // Fix #22: Cache hash results to avoid repeated SHA256
    private var hashCache: [String: String] = [:]

    // Fix #12: In-memory LRU tracking instead of filesystem writes
    private var accessTimes: [String: Date] = [:]

    // In-memory image cache (NSCache auto-evicts under memory pressure)
    // This prevents re-reading from disk when LazyVStack recycles views
    private let memoryCache = NSCache<NSString, UIImage>()

    private init() {
        let cacheBase = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
        cacheDirectory = cacheBase.appendingPathComponent("Images", isDirectory: true)

        try? FileManager.default.createDirectory(
            at: cacheDirectory, withIntermediateDirectories: true)

        // NSCache auto-evicts under memory pressure — no artificial limit needed
        // All loaded images stay in RAM for instant access until iOS reclaims memory

        // Build in-memory file index on init
        Task { await buildIndex() }
    }

    // MARK: - Public API

    /// Load image from cache (memory → disk)
    /// Returns instantly from memory if previously loaded this session
    func loadImage(for urlString: String) -> UIImage? {
        let hash = cachedHash(urlString)
        let cacheKey = hash as NSString

        // Layer 1: Check in-memory cache (instant, no disk I/O)
        if let memImage = memoryCache.object(forKey: cacheKey) {
            // Fix #12: Track access time in memory, not filesystem
            accessTimes[hash] = Date()
            return memImage
        }

        // Layer 2: Check disk
        guard fileIndex.contains(hash) else { return nil }

        let filePath = cacheDirectory.appendingPathComponent(hash)
        guard let data = try? Data(contentsOf: filePath),
            let image = UIImage(data: data)
        else {
            fileIndex.remove(hash)
            return nil
        }

        // Promote to memory cache for next access
        memoryCache.setObject(image, forKey: cacheKey)

        // Fix #12: Track access in-memory (no filesystem write)
        accessTimes[hash] = Date()

        return image
    }

    /// Save image to disk cache
    func saveImage(_ image: UIImage, for urlString: String) {
        let hash = cachedHash(urlString)
        let cacheKey = hash as NSString

        // Save to memory cache immediately
        memoryCache.setObject(image, forKey: cacheKey)

        // Use JPEG for photos, PNG for images with transparency
        guard let data = image.jpegData(compressionQuality: 0.85) else { return }

        let filePath = cacheDirectory.appendingPathComponent(hash)
        try? data.write(to: filePath, options: .atomic)

        fileIndex.insert(hash)
        accessTimes[hash] = Date()
        // Fix #13: Track size incrementally
        currentCacheSize += Int64(data.count)

        // Trigger eviction check in background (only if over limit)
        if currentCacheSize > maxCacheSize {
            Task.detached(priority: .utility) {
                await self.evictIfNeeded()
            }
        }
    }

    /// Save raw image data to disk cache (avoids re-encoding)
    func saveData(_ data: Data, for urlString: String) {
        let hash = cachedHash(urlString)
        let cacheKey = hash as NSString

        let filePath = cacheDirectory.appendingPathComponent(hash)
        try? data.write(to: filePath, options: .atomic)

        fileIndex.insert(hash)
        accessTimes[hash] = Date()
        // Fix #13: Track size incrementally
        currentCacheSize += Int64(data.count)

        // Also promote to memory cache
        if let image = UIImage(data: data) {
            memoryCache.setObject(image, forKey: cacheKey)
        }
    }

    /// Check if image exists in disk cache
    func hasImage(for urlString: String) -> Bool {
        return fileIndex.contains(cachedHash(urlString))
    }

    /// Prefetch multiple image URLs in parallel
    /// Fix #23: Prefetch with limited concurrency (max 6 concurrent downloads)
    func prefetch(urls: [String]) async {
        let maxConcurrent = 6
        await withTaskGroup(of: Void.self) { group in
            var active = 0
            for urlString in urls {
                // Skip if already cached
                guard !hasImage(for: urlString) else { continue }
                guard let url = URL(string: urlString) else { continue }

                if active >= maxConcurrent {
                    await group.next()  // Wait for one to finish
                    active -= 1
                }

                active += 1
                group.addTask {
                    do {
                        let request = URLRequest(
                            url: url,
                            cachePolicy: .returnCacheDataElseLoad,
                            timeoutInterval: 30
                        )
                        let (data, response) = try await URLSession.shared.data(for: request)

                        if let httpResponse = response as? HTTPURLResponse,
                            httpResponse.statusCode == 200,
                            UIImage(data: data) != nil
                        {
                            await self.saveData(data, for: urlString)
                        }
                    } catch {
                        // Silent fail for prefetch
                    }
                }
            }
        }
    }

    /// Clear all cached images
    func clearAll() {
        let files =
            (try? fileManager.contentsOfDirectory(
                at: cacheDirectory, includingPropertiesForKeys: nil)) ?? []
        for file in files {
            try? fileManager.removeItem(at: file)
        }
        fileIndex.removeAll()
    }

    /// Get cache stats
    func getStats() -> (fileCount: Int, totalSize: Int64) {
        var totalSize: Int64 = 0
        let files =
            (try? fileManager.contentsOfDirectory(
                at: cacheDirectory, includingPropertiesForKeys: [.fileSizeKey])) ?? []
        for file in files {
            let attrs = try? file.resourceValues(forKeys: [.fileSizeKey])
            totalSize += Int64(attrs?.fileSize ?? 0)
        }
        return (files.count, totalSize)
    }

    // MARK: - Private Helpers

    private func buildIndex() {
        guard
            let files = try? fileManager.contentsOfDirectory(
                at: cacheDirectory, includingPropertiesForKeys: [.fileSizeKey])
        else { return }
        currentCacheSize = 0
        for file in files {
            fileIndex.insert(file.lastPathComponent)
            let size = (try? file.resourceValues(forKeys: [.fileSizeKey]))?.fileSize ?? 0
            currentCacheSize += Int64(size)
        }
        AppLogger.debug(
            "[ImageDiskCache] Indexed \(fileIndex.count) cached images (\(currentCacheSize / 1024 / 1024)MB)"
        )
    }

    // Fix #13: Uses in-memory accessTimes for LRU instead of re-scanning directory
    private func evictIfNeeded() {
        guard currentCacheSize > maxCacheSize else { return }

        // Sort by in-memory access times (oldest first)
        let sortedHashes = accessTimes.sorted { $0.value < $1.value }.map { $0.key }

        let targetSize = maxCacheSize / 2

        for hash in sortedHashes {
            guard currentCacheSize > targetSize else { break }
            let filePath = cacheDirectory.appendingPathComponent(hash)
            let size = (try? filePath.resourceValues(forKeys: [.fileSizeKey]))?.fileSize ?? 0
            try? fileManager.removeItem(at: filePath)
            fileIndex.remove(hash)
            accessTimes.removeValue(forKey: hash)
            currentCacheSize -= Int64(size)
        }

        AppLogger.debug("[ImageDiskCache] Evicted to \(currentCacheSize / 1024 / 1024)MB")
    }

    // Fix #22: Cache SHA256 results to avoid repeated computation
    private func cachedHash(_ string: String) -> String {
        if let cached = hashCache[string] { return cached }
        let inputData = Data(string.utf8)
        let hashed = SHA256.hash(data: inputData)
        let result = hashed.compactMap { String(format: "%02x", $0) }.joined()
        hashCache[string] = result
        return result
    }
}

File: apps/Applications/android/app/src/main/java/com/ecommerceearn/app/data/manager/ImageDiskCache.kt
package com.ecommerceearn.app.data.manager

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.LruCache
import com.ecommerceearn.app.utils.AppLogger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import java.io.File
import java.io.FileOutputStream
import java.net.URL
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap
import java.net.HttpURLConnection

object ImageDiskCache {
    private const val MAX_CACHE_SIZE = 500L * 1024L * 1024L // 500 MB
    private lateinit var cacheDir: File

    private val fileIndex = ConcurrentHashMap.newKeySet<String>()
    private var currentCacheSize: Long = 0L
    private val hashCache = ConcurrentHashMap<String, String>()
    private val accessTimes = ConcurrentHashMap<String, Long>()

    // Memory cache limits size in KB
    private val memoryCache = object : LruCache<String, Bitmap>( (Runtime.getRuntime().maxMemory() / 1024 / 8).toInt() ) {
        override fun sizeOf(key: String, bitmap: Bitmap): Int {
            return bitmap.byteCount / 1024
        }
    }

    fun init(context: Context) {
        cacheDir = File(context.cacheDir, "Images").apply { mkdirs() }

        CoroutineScope(Dispatchers.IO).launch {
            buildIndex()
        }
    }

    suspend fun loadImage(urlString: String): Bitmap? = withContext(Dispatchers.IO) {
        val hash = cachedHash(urlString)

        memoryCache.get(hash)?.let { memImage ->
            accessTimes[hash] = System.currentTimeMillis()
            return@withContext memImage
        }

        if (!fileIndex.contains(hash)) return@withContext null

        val filePath = File(cacheDir, hash)
        if (filePath.exists()) {
            try {
                val bitmap = BitmapFactory.decodeFile(filePath.absolutePath)
                if (bitmap != null) {
                    memoryCache.put(hash, bitmap)
                    accessTimes[hash] = System.currentTimeMillis()
                    return@withContext bitmap
                } else {
                    fileIndex.remove(hash)
                }
            } catch (e: Exception) {
                fileIndex.remove(hash)
            }
        }
        return@withContext null
    }

    suspend fun saveImage(bitmap: Bitmap, urlString: String) = withContext(Dispatchers.IO) {
        val hash = cachedHash(urlString)
        memoryCache.put(hash, bitmap)

        val filePath = File(cacheDir, hash)
        try {
            FileOutputStream(filePath).use { out ->
                // Compress logic (approximating iOS .jpegData(quality: 0.85))
                bitmap.compress(Bitmap.CompressFormat.JPEG, 85, out)
            }
            
            val size = filePath.length()
            fileIndex.add(hash)
            accessTimes[hash] = System.currentTimeMillis()
            currentCacheSize += size

            if (currentCacheSize > MAX_CACHE_SIZE) {
                evictIfNeeded()
            }
        } catch (e: Exception) {
            AppLogger.error("[ImageDiskCache] Failed to save image: ${e.message}")
        }
    }

    suspend fun saveData(data: ByteArray, urlString: String) = withContext(Dispatchers.IO) {
        val hash = cachedHash(urlString)
        val filePath = File(cacheDir, hash)

        try {
            filePath.writeBytes(data)
            val size = data.size.toLong()
            fileIndex.add(hash)
            accessTimes[hash] = System.currentTimeMillis()
            currentCacheSize += size

            val bitmap = BitmapFactory.decodeByteArray(data, 0, data.size)
            if (bitmap != null) {
                memoryCache.put(hash, bitmap)
            }

            if (currentCacheSize > MAX_CACHE_SIZE) {
                evictIfNeeded()
            }
        } catch (e: Exception) {
            AppLogger.error("[ImageDiskCache] Failed to save raw data: ${e.message}")
        }
    }

    fun hasImage(urlString: String): Boolean {
        return fileIndex.contains(cachedHash(urlString))
    }

    suspend fun prefetch(urls: List<String>) = withContext(Dispatchers.IO) {
        // Implement concurrent limited downloads via coroutines
        val defs = urls.filter { !hasImage(it) }.map { urlString ->
            async {
                try {
                    val url = URL(urlString)
                    val connection = url.openConnection() as HttpURLConnection
                    connection.connectTimeout = 30000
                    connection.readTimeout = 30000
                    
                    if (connection.responseCode == 200) {
                        val bytes = connection.inputStream.readBytes()
                        saveData(bytes, urlString)
                    }
                } catch (e: Exception) {
                    // Silent fail for prefetch
                }
            }
        }
        defs.awaitAll()
    }

    fun clearAll() {
        cacheDir.listFiles()?.forEach { it.delete() }
        memoryCache.evictAll()
        fileIndex.clear()
        currentCacheSize = 0L
        accessTimes.clear()
    }

    private fun buildIndex() {
        val files = cacheDir.listFiles() ?: return
        currentCacheSize = 0L
        for (file in files) {
            fileIndex.add(file.name)
            currentCacheSize += file.length()
        }
        AppLogger.debug("[ImageDiskCache] Indexed ${fileIndex.size} cached images (${currentCacheSize / 1024 / 1024}MB)")
    }

    private fun evictIfNeeded() {
        if (currentCacheSize <= MAX_CACHE_SIZE) return

        val sortedHashes = accessTimes.toList()
            .sortedBy { (_, time) -> time }
            .map { it.first }

        val targetSize = MAX_CACHE_SIZE / 2

        for (hash in sortedHashes) {
            if (currentCacheSize <= targetSize) break
            val filePath = File(cacheDir, hash)
            if (filePath.exists()) {
                val size = filePath.length()
                if (filePath.delete()) {
                    fileIndex.remove(hash)
                    accessTimes.remove(hash)
                    currentCacheSize -= size
                }
            }
        }
        AppLogger.debug("[ImageDiskCache] Evicted to ${currentCacheSize / 1024 / 1024}MB")
    }

    private fun cachedHash(string: String): String {
        return hashCache.getOrPut(string) {
            val digest = MessageDigest.getInstance("SHA-256")
            val hash = digest.digest(string.toByteArray())
            hash.joinToString("") { "%02x".format(it) }
        }
    }
}
